File: .test/ziwei-test/lib/ziwei/utils/calc_commons.rb
module Ziwei
  module Utils
    module CalcCommons
      module ClassMethods
      end
      
      module InstanceMethods
        def calc_self_body_position(birth_hour, birth_month)
          hour_index = Constants::Branches::Indexes[birth_hour]
          month_index = Constants::Branches::Indexes[birth_month]

          self_index = limit_dec(month_index, 12, hour_index-1)
          body_index = limit_inc(month_index, 12, hour_index-1)

          [
            Constants::Branches::Orders[self_index],
            Constants::Branches::Orders[body_index]
          ]
        end

        def calc_palaces_positions(self_position)
          self_index = Constants::Branches::Indexes[self_position]
          
          branches = Constants::Branches::Names.keys
          palaces = Constants::Palaces::Orders

          merge_sequences_from_index(branches, palaces, self_index)
        end

        def calc_cuc(self_position, birth_year_stem)
          start_element = Constants::Wuxing::ElementsByBranches[self_position]
          start_element_index = Constants::Wuxing::Elements.index(start_element)
          counting_steps = Constants::Stems::Indexes[birth_year_stem]
          cuc_index = limit_inc(start_element_index, 5, counting_steps, 0)

          cuc_element = Constants::Wuxing::Elements[cuc_index]
          cuc_number = Constants::Wuxing::CucByElements[cuc_element]

          [cuc_element, cuc_number]
        end

        def merge_sequences_from_index(first_sequence, second_sequence, merge_index, reversed_order = false)
          second_seq = second_sequence.dup

          # re-arrange second sequence in order of first one,
          # skip if merge_index is 1 (2 sequences have identical order)
          if merge_index != 1
            second_seq = unless reversed_order
              second_seq[(12-merge_index+1)..-1] + second_seq[0..(12-merge_index)]
            else
              second_seq[0..merge_index-1].reverse + second_seq[merge_index..-1].reverse
            end
          end

          Hash[[first_sequence, second_seq].transpose]
        end
      end
      
      def self.included(receiver)
        receiver.extend         ClassMethods
        receiver.send :include, InstanceMethods
      end
    end
  end
end

File: ziwei-js/src/utils/calc_commons.js
_Ziwei_Calculator.prototype.calcSelfBodyPosition = function(birthHour, birthMonth) {
  var hourIndex = Ziwei.Configs.Branches.Indexes[birthHour];
  var monthIndex = Ziwei.Configs.Branches.Indexes[birthMonth];

  var selfIndex = monthIndex.limitInc(-hourIndex+1);
  var bodyIndex = monthIndex.limitInc(hourIndex-1);

  return [
    Ziwei.Configs.Branches.Orders[selfIndex],
    Ziwei.Configs.Branches.Orders[bodyIndex]
  ];
};

_Ziwei_Calculator.prototype.calcCuc = function(selfPosition, birthYearStem) {
  var startElement = Ziwei.Configs.Wuxing.ElementsByBranches[selfPosition];
  var startElementIndex = Ziwei.Configs.Wuxing.Elements.indexOf(startElement);
  
  var countingSteps = Ziwei.Configs.Stems.Indexes[birthYearStem];
  var cucIndex = startElementIndex.limitInc(countingSteps, 5, 0);

  var cucElement = Ziwei.Configs.Wuxing.Elements[cucIndex];
  var cucNumber = Ziwei.Configs.Wuxing.CucByElements[cucElement];

  return [cucElement, cucNumber];
};

_Ziwei_Calculator.prototype.calcPalacesPositions = function(selfPosition) {
  var selfIndex = Ziwei.Configs.Branches.Indexes[selfPosition];
  var branches = Ziwei.Configs.Branches.Names.allKeys();
  var palaces = Ziwei.Configs.Palaces.Orders;

  return this.mergeSequencesFromIndex(branches, palaces, selfIndex);
};

_Ziwei_Calculator.prototype.calcOpportunityAges = function(selfPosition, cucNumber, fateDirection) {
  var selfIndex = Ziwei.Configs.Branches.Indexes[selfPosition];
  var branches = Ziwei.Configs.Branches.Names.allKeys();
  var ages = Array.fromRange(0, 11).map((mult) => cucNumber + 10*mult);
  
  return this.mergeSequencesFromIndex(branches, ages, selfIndex, fateDirection === -1);
};

_Ziwei_Calculator.prototype.calcConnectedPalaceCoordinates = function(selfPosition) {
  var selfIndex = Ziwei.Configs.Branches.Indexes[selfPosition];
  var selfCoordinate = Ziwei.Configs.Palaces.DrawingRootCoordinates[selfPosition];

  var oppositeIndex = selfIndex.limitInc(6);
  var oppositePosition = Ziwei.Configs.Branches.Orders[oppositeIndex];
  var oppositeCoordinate = Ziwei.Configs.Palaces.DrawingRootCoordinates[oppositePosition];

  var trilogyElement = Ziwei.Configs.BranchSets.Trilogy.ByBranches[selfPosition];
  var sameSetPositions = Ziwei.Configs.BranchSets.Trilogy.ByElements[trilogyElement].slice();
  var selfPositionIndex = selfPosition.indexOf(selfPosition);
  sameSetPositions = sameSetPositions.filter((x, i) => i !== selfPositionIndex);

  var sameSetCoordinates = sameSetPositions.map((position) =>
    Ziwei.Configs.Palaces.DrawingRootCoordinates[position]
  );

  var coordinates = [oppositeCoordinate];
  coordinates.push(...sameSetCoordinates);

  return coordinates.map((fromCoordinate) => [fromCoordinate, selfCoordinate]);
};